cap_build.c

```diff
@@ -0,0 +1,68 @@
+#include "cheriintrin.h"
+#include "include/common.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+int main()
+{
+    // We create two capabilities: x, and y, which is derived from x. We note that, though we use
+    // `cheri_address_set()`, it doesn't actually set (i.e. _change_) the address of its arguments,
+    // but construct a new capability with the address we pass.
+    //
+    // cheri_address_set(void* __capability, uintptr_t) -> void *__capability
+    void *__capability x = cheri_ddc_get();
+    uintptr_t y_ptr = (uintptr_t) malloc(20);
+    void *__capability y = cheri_address_set(x, y_ptr);
+
+    printf("\n\tThe second capability printed below is created from the first using");
+    printf(
+        "\n\t`cheri_address_set()`. Notice that `cheri_address_set()` created a new capability,");
+    printf("\n\trather than setting the address on the first.");
+    printf("\n\n");
+    pp_cap(x);
+    pp_cap(y);
+    printf("\n");
+
+    // We create an (invalid!) capability by writing it directly.
+    void *my_cap_mem = malloc(16);
+    long cap;
+    cap = 0xdc5f400000030005;
+    memcpy(my_cap_mem + 8, &cap, 8);
+    cap = 0x00000000deadbeef;
+    memcpy(my_cap_mem + 0, &cap, 8);
+    void *__capability my_cap = (void *__capability) *(void *__capability *) my_cap_mem;
+
+    // We build a new capability using `cheri_cap_build()`. Note that this capability is _valid_,
+    // because it's derived from `x`, which is valid. The address comes from `my_cap`, which isn't
+    // valid, but we can still create a valid capability beecause `cheri_cap_build` just reads it's
+    // address bits and constructs a capability.
+
+    // cheri_cap_build(void *__capability, uintcap_t) -> void *__capability
+    void *__capability valid_cap = cheri_cap_build(x, (uintcap_t) my_cap);
```

Comment:
Can you explain why we need the `uintcap_t` cast here? I *think* it's because it's 128 bits, rather than a capability's 129, but not sure. You can also just say we need this since it works.

Reply:
We need this since it works 😛

More seriously though, the function only accepts a `uintcap_t`. You might remember when we were pairing on this that we got lots of cryptic errors from the compiler around the type of this argument. It wanted a `uintcap_t` — I imagine because it specifically doesn't care about the validity bit of its second argument (as it'll decide for itself whether the capability it returns should be valid), but it needs every other bit of the capability, because whatever it returns will be a capability with these 128 bits…just not necessarily a valid one.

Reply:
I meant more in the code, as a comment. I recall we found some builtin function signature of this `cheri_cap_build` that took two identically typed arguments in the representation, but passing a capability as the second one gave an error (while obviously a capability as the first one is fine). 

If you, by chance, found some more details about the function signature of this, that would be great, if not, no worries. Can say we experimented and this seems to be the signature, just add that as a comment.

Reply:
Ah! Sorry, misunderstood. Closest to a definition I ever found was at the end of [this article from ARM](https://www.morello-project.org/resources/morello-linux-morelloie/). Added a note on that in [f0dbba6](https://github.com/capablevms/cheri-examples/pull/92/commits/f0dbba61699584ac1fa59191d2d7c276e2662669).

Reply:
Right, Yury did mention when that article got uploaded. I wonder how related it is to his examples repo.